specifications/tcp/tcp.tla

```diff
@@ -0,0 +1,330 @@
+
+CLOSE_SYN_SENT(local, remote) ==
+    \* User command: CLOSE
+    /\ local # remote
+    /\ connstate[local] = "SYN-SENT"
+    /\ UNCHANGED network
+    /\ connstate' = [connstate EXCEPT ![local] = "CLOSED"]
+    /\ tcb' = [tcb EXCEPT ![local] = FALSE]
+
+CLOSE_SYN_RECEIVED(local, remote) ==
+    \* User Command: CLOSE
+    /\ local # remote
+    /\ connstate[local] = "SYN-RECEIVED"
+    /\ network' = [ network EXCEPT ![remote] = Append(@, "FIN")]
+    /\ connstate' = [connstate EXCEPT ![local] = "FIN-WAIT-1"]
+    /\ UNCHANGED tcb
+
+SEND(local, remote) ==
+    \* User command: SEND
+    /\ local # remote
+    /\ connstate[local] = "LISTEN"
+    /\ network' = [ network EXCEPT ![remote] = Append(@, "SYN")]
+    /\ connstate' = [connstate EXCEPT ![local] = "SYN-SENT"]
+    /\ UNCHANGED tcb
+
+CLOSE_LISTEN(local, remote) ==
+    \* User command: CLOSE
+    /\ local # remote
+    /\ connstate[local] = "LISTEN"
+    /\ UNCHANGED network
+    /\ connstate' = [connstate EXCEPT ![local] = "CLOSED"]
+    /\ tcb' = [tcb EXCEPT ![local] = FALSE]
+
+CLOSE_ESTABLISHED(local, remote) ==
+    \* User command: CLOSE
+    /\ local # remote
+    /\ connstate[local] = "ESTABLISHED"
+    /\ network' = [ network EXCEPT ![remote] = Append(@, "FIN")]
+    /\ connstate' = [connstate EXCEPT ![local] = "FIN-WAIT-1"]
+    /\ UNCHANGED tcb
+
+CLOSE_CLOSE_WAIT(local, remote) ==
+    \* User command: CLOSE
+    /\ local # remote
+    /\ connstate[local] = "CLOSE-WAIT"
+    /\ network' = [ network EXCEPT ![remote] = Append(@, "FIN")]
+    /\ connstate' = [connstate EXCEPT ![local] = "LAST-ACK"]
+    /\ UNCHANGED tcb
+
+User ==
+    \E local, remote \in Peers :
+        \/ ACTIVE_OPEN(local, remote)
+        \/ PASSIVE_OPEN(local, remote)
+        \/ CLOSE_SYN_SENT(local, remote)
+        \/ CLOSE_SYN_RECEIVED(local, remote)
+        \/ CLOSE_LISTEN(local, remote)
+        \/ CLOSE_ESTABLISHED(local, remote)
+        \/ CLOSE_CLOSE_WAIT(local, remote)
+        \/ SEND(local, remote)
+
+-----------------------------------------------------------------------------
+       
+SynSent(local, remote) ==
+    /\ local # remote
+    /\ connstate[local] = "SYN-SENT"
+    /\ UNCHANGED tcb
+    /\ \/ /\ IsPrefix(<<"SYN">>, network[local])
+          /\ network' = [ network EXCEPT ![remote] = Append(@, "SYN,ACK"),
+                                         ![local] = Tail(network[local])]
+          /\ connstate' = [connstate EXCEPT ![local] = "SYN-RECEIVED"]
+       \/ /\ IsPrefix(<<"SYN,ACK">>, network[local])
+          /\ network' = [ network EXCEPT ![remote] = Append(@, "ACK"),
+                                         ![local] = Tail(network[local])]
+          /\ connstate' = [connstate EXCEPT ![local] = "ESTABLISHED"]
+
+SynReceived(local, remote) ==
+    /\ local # remote
+    /\ connstate[local] = "SYN-RECEIVED"
+    /\ UNCHANGED tcb
+    /\ \/ /\ IsPrefix(<<"RST">>, network[local]) \* TODO: note 1 in RFC 9293 (must have gotten here from a *passive* OPEN, i.e., LISTEN)
```

Comment:
Indeed, I am not sure why the condition in this note is important. Did you get it?

Reply:
More states (1664/8952) but no other difference when modeled explicitly:

```diff
diff --git a/specifications/tcp/MCtcp.tla b/specifications/tcp/MCtcp.tla
index e0cee03..fb64c8e 100644
--- a/specifications/tcp/MCtcp.tla
+++ b/specifications/tcp/MCtcp.tla
@@ -12,6 +12,7 @@ MCInit ==
         /\ tcb = [p \in Peers |-> TRUE]
         /\ connstate = [p \in Peers |-> IF p = node THEN "SYN-SENT" ELSE "LISTEN"]
         /\ network = [p \in Peers |-> IF p = node THEN <<>> ELSE << "SYN" >>]
+        /\ listened = [p \in Peers |-> IF connstate[p] = "LISTEN" THEN TRUE ELSE FALSE]
 
 NoRetransmission ==
     \A p \in Peers :
diff --git a/specifications/tcp/tcp.tla b/specifications/tcp/tcp.tla
index ae312f6..57329aa 100644
--- a/specifications/tcp/tcp.tla
+++ b/specifications/tcp/tcp.tla
@@ -19,18 +19,21 @@ States == {"LISTEN", "CLOSED", "SYN-SENT", "SYN-RECEIVED", "ESTABLISHED",
 
 VARIABLE
     tcb,
+    wasListening,
     connstate,
     network
 
-vars == <<tcb, connstate, network>>
+vars == <<tcb, wasListening, connstate, network>>
 
 TypeOK ==
     /\ tcb \in [ Peers -> BOOLEAN ]
+    /\ wasListening \in [ Peers -> BOOLEAN ]
     /\ connstate \in [ Peers -> States ]
     /\ network \in [ Peers -> Seq({"SYN", "SYN,ACK", "ACK", "RST", "FIN"} \cup {"ACKofFIN"}) ]
 
 Init ==
     /\ tcb = [p \in Peers |-> FALSE]
+    /\ wasListening = [p \in Peers |-> FALSE]
     /\ connstate = [p \in Peers |-> "CLOSED"]
     /\ network = [p \in Peers |-> <<>>]
 
@@ -43,6 +46,7 @@ PASSIVE_OPEN(local, remote) ==
     /\ UNCHANGED network
     /\ connstate' = [connstate EXCEPT ![local] = "LISTEN"]
     /\ tcb' = [tcb EXCEPT ![local] = TRUE]
+    /\ wasListening' = [wasListening EXCEPT ![local] = TRUE]
 
 ACTIVE_OPEN(local, remote) ==
     \* User command: ACTIVE-OPEN
@@ -51,6 +55,7 @@ ACTIVE_OPEN(local, remote) ==
     /\ network' = [ network EXCEPT ![remote] = Append(@, "SYN")]
     /\ connstate' = [connstate EXCEPT ![local] = "SYN-SENT"]
     /\ tcb' = [tcb EXCEPT ![local] = TRUE]
+    /\ wasListening' = [wasListening EXCEPT ![local] = TRUE]
 
 CLOSE_SYN_SENT(local, remote) ==
     \* User command: CLOSE
@@ -59,6 +64,7 @@ CLOSE_SYN_SENT(local, remote) ==
     /\ UNCHANGED network
     /\ connstate' = [connstate EXCEPT ![local] = "CLOSED"]
     /\ tcb' = [tcb EXCEPT ![local] = FALSE]
+    /\ UNCHANGED wasListening
 
 CLOSE_SYN_RECEIVED(local, remote) ==
     \* User Command: CLOSE
@@ -67,6 +73,7 @@ CLOSE_SYN_RECEIVED(local, remote) ==
     /\ network' = [ network EXCEPT ![remote] = Append(@, "FIN")]
     /\ connstate' = [connstate EXCEPT ![local] = "FIN-WAIT-1"]
     /\ UNCHANGED tcb
+    /\ UNCHANGED wasListening
 
 SEND(local, remote) ==
     \* User command: SEND
@@ -75,6 +82,7 @@ SEND(local, remote) ==
     /\ network' = [ network EXCEPT ![remote] = Append(@, "SYN")]
     /\ connstate' = [connstate EXCEPT ![local] = "SYN-SENT"]
     /\ UNCHANGED tcb
+    /\ UNCHANGED wasListening
 
 CLOSE_LISTEN(local, remote) ==
     \* User command: CLOSE
@@ -83,6 +91,7 @@ CLOSE_LISTEN(local, remote) ==
     /\ UNCHANGED network
     /\ connstate' = [connstate EXCEPT ![local] = "CLOSED"]
     /\ tcb' = [tcb EXCEPT ![local] = FALSE]
+    /\ UNCHANGED wasListening
 
 CLOSE_ESTABLISHED(local, remote) ==
     \* User command: CLOSE
@@ -91,6 +100,7 @@ CLOSE_ESTABLISHED(local, remote) ==
     /\ network' = [ network EXCEPT ![remote] = Append(@, "FIN")]
     /\ connstate' = [connstate EXCEPT ![local] = "FIN-WAIT-1"]
     /\ UNCHANGED tcb
+    /\ UNCHANGED wasListening
 
 CLOSE_CLOSE_WAIT(local, remote) ==
     \* User command: CLOSE
@@ -99,6 +109,7 @@ CLOSE_CLOSE_WAIT(local, remote) ==
     /\ network' = [ network EXCEPT ![remote] = Append(@, "FIN")]
     /\ connstate' = [connstate EXCEPT ![local] = "LAST-ACK"]
     /\ UNCHANGED tcb
+    /\ UNCHANGED wasListening
 
 User ==
     \E local, remote \in Peers :
@@ -117,6 +128,7 @@ SynSent(local, remote) ==
     /\ local # remote
     /\ connstate[local] = "SYN-SENT"
     /\ UNCHANGED tcb
+    /\ UNCHANGED wasListening
     /\ \/ /\ IsPrefix(<<"SYN">>, network[local])
           /\ network' = [ network EXCEPT ![remote] = Append(@, "SYN,ACK"),
                                          ![local] = Tail(network[local])]
@@ -130,7 +142,9 @@ SynReceived(local, remote) ==
     /\ local # remote
     /\ connstate[local] = "SYN-RECEIVED"
     /\ UNCHANGED tcb
-    /\ \/ /\ IsPrefix(<<"RST">>, network[local]) \* TODO: note 1 in RFC 9293 (must have gotten here from a *passive* OPEN, i.e., LISTEN)
+    /\ UNCHANGED wasListening
+    /\ \/ /\ wasListening[local]
+          /\ IsPrefix(<<"RST">>, network[local]) \* TODO: note 1 in RFC 9293 (must have gotten here from a *passive* OPEN, i.e., LISTEN)
           /\ network' = [ network EXCEPT ![local] = Tail(network[local])]
           /\ connstate' = [connstate EXCEPT ![local] = "LISTEN"]
        \/ /\ IsPrefix(<<"ACK">>, network[local])
@@ -141,6 +155,7 @@ Listen(local, remote) ==
     /\ local # remote
     /\ connstate[local] = "LISTEN"
     /\ UNCHANGED tcb
+    /\ UNCHANGED wasListening
     /\ \/ /\ IsPrefix(<<"SYN">>, network[local])
           /\ network' = [ network EXCEPT ![remote] = Append(@, "SYN,ACK"),
                                          ![local] = Tail(network[local])]
@@ -150,6 +165,7 @@ Established(local, remote) ==
     /\ local # remote
     /\ connstate[local] = "ESTABLISHED"
     /\ UNCHANGED tcb
+    /\ UNCHANGED wasListening
     /\ IsPrefix(<<"FIN">>, network[local])
     /\ network' = [ network EXCEPT ![remote] = Append(@, "ACKofFIN"),
                                    ![local] = Tail(network[local])]
@@ -161,6 +177,7 @@ FinWait1(local, remote) ==
     /\ local # remote
     /\ connstate[local] = "FIN-WAIT-1"
     /\ UNCHANGED tcb
+    /\ UNCHANGED wasListening
     /\ \/ /\ IsPrefix(<<"FIN">>, network[local])
           /\ network' = [ network EXCEPT ![remote] = Append(@, "ACKofFIN"),
                                          ![local] = Tail(network[local])]
@@ -177,6 +194,7 @@ FinWait2(local, remote) ==
     /\ network' = [ network EXCEPT ![remote] = Append(@, "ACKofFIN"),
                                    ![local] = Tail(network[local])]
     /\ connstate' = [connstate EXCEPT ![local] = "TIME-WAIT"]
+    /\ UNCHANGED wasListening
 
 Closing(local, remote) ==
     /\ local # remote
@@ -185,6 +203,7 @@ Closing(local, remote) ==
     /\ IsPrefix(<<"ACKofFIN">>, network[local])
     /\ network' = [ network EXCEPT ![local] = Tail(network[local])]
     /\ connstate' = [connstate EXCEPT ![local] = "TIME-WAIT"]
+    /\ UNCHANGED wasListening
 
 LastAck(local, remote) ==
     /\ local # remote
@@ -193,6 +212,7 @@ LastAck(local, remote) ==
     /\ IsPrefix(<<"ACKofFIN">>, network[local])
     /\ network' = [ network EXCEPT ![local] = Tail(network[local])]
     /\ connstate' = [connstate EXCEPT ![local] = "CLOSED"]
+    /\ UNCHANGED wasListening
 
 TimeWait(local, remote) ==
     /\ local # remote
@@ -201,6 +221,7 @@ TimeWait(local, remote) ==
     /\ UNCHANGED network
     /\ connstate' = [connstate EXCEPT ![local] = "CLOSED"]
     /\ tcb' = [tcb EXCEPT ![local] = FALSE]
+    /\ UNCHANGED wasListening
     \* TODO: Timeout=2MSL
 
 Note2(local, remote) ==
@@ -208,7 +229,7 @@ Note2(local, remote) ==
        and the local FIN is also acknowledged. *)
     /\ local # remote
     /\ connstate[local] = "FIN-WAIT-1"
-    /\ UNCHANGED tcb
+    /\ UNCHANGED <<tcb, wasListening>>
     \* The local FIN is also acknowledged, i.e., skipping the CLOSING state.
     /\ IsPrefix(<< "FIN", "ACKofFIN" >>, network[local])
     \* ??? The RFC doesn't explictly mention this, but we send the ACKofFIN here.
@@ -237,7 +258,7 @@ Note3(local, remote) ==
     \* ??? could also be interpreted as those states from where we can tranisiton to TIME-WAIT,
     \* ??? i.e., FIN-WAIT-1, FIN-WAIT-2, CLOSING.
     /\ local # remote
-    /\ UNCHANGED tcb
+    /\ UNCHANGED <<tcb, wasListening>>
     /\ \/ /\ tcb[local] \* ??? RFC 9293 doesn't mention this, but it seems impossible to send a RST without a TCB.
           /\ network' = [ network EXCEPT ![remote] = Append(@, "RST")]
           /\ connstate' = [connstate EXCEPT ![local] = "TIME-WAIT"]
```

Going to leave the TODO in hopes that somebody with more knowledge about TCP will clarify eventually.